cli/pcluster/pcluster.py

```diff
@@ -595,12 +633,28 @@ def command(args, extra_args):
         stack_result = cfn.describe_stacks(StackName=stack).get("Stacks")[0]
         status = stack_result.get("StackStatus")
         valid_status = ["CREATE_COMPLETE", "UPDATE_COMPLETE", "UPDATE_ROLLBACK_COMPLETE"]
-        if status not in valid_status:
-            LOGGER.info("Stack status: %s. Stack needs to be in %s", status, " or ".join(valid_status))
+        invalid_status = ["DELETE_COMPLETE", "DELETE_IN_PROGRESS"]
+
+        if status in invalid_status:
+            LOGGER.info("Stack status: %s. Cannot SSH while in %s" % (status, " or ".join(invalid_status)))
             sys.exit(1)
-        outputs = stack_result.get("Outputs")
-        username = [o.get("OutputValue") for o in outputs if o.get("OutputKey") == "ClusterUser"][0]
-        ip = [o.get("OutputValue") for o in outputs if o.get("OutputKey") == "MasterPublicIP"][0]
+        elif status in valid_status:
+            outputs = stack_result.get("Outputs")
+            username = _get_output_value(outputs, "ClusterUser")
+
+            try:
+                ip = _get_output_value(outputs, "MasterPublicIP")
+            except IndexError:
+                # If they're not using public ip's use the private ip instead
+                ip = _get_output_value(outputs, "MasterPrivateIP")
+        else:
+            # Stack is in CREATING, CREATED_FAILED, or ROLLBACK_COMPLETE but MasterServer is running
+            ip = _get_master_server_ip(stack, config)
+            stack_result = cfn.describe_stacks(StackName=stack).get("Stacks")[0]
```

Comment:
Can we reuse the stack_result we already retrieved above?

Reply:
yes, changed